### PR TITLE
Allow a blacklist to only apply to a url

### LINKED ISSUE #8: Can't add blacklist entries per repository
I'm looking for a way to add blacklist entries per repository. 

The concrete problem is I have both of alienBOB's repositories configured with slackroll (restricted_sbrepos and sbrepos) and want to blacklist some packages in sborepos which are also in restricted_sbrepos so I am not prompted to choose each time a package is updated (like vlc and ffmpeg).

Since the package names and paths are exactly the same, the normal blacklist mechanism doesn't work.

A second use case would be only wanting to install a couple of packages from a repo and not wanting to be prompted about new packages every time a package is added. e.g. only wanting to install chromium and palemoon from alienBOB's repo and not caring about anything else in it. One solution would be setting the blacklist for that repo to `^\.\/(?!chromium|palemoon).*$`

I have started working on a solution but would like to confirm the approach before I prepare a PR.

I don't want to break anyone's existing blacklist, so I was thinking of this:

Change `get_blacklist()` to return a dict, keyed by repo id (None or some non-digit value will be the key for the primary mirror) and values which are arrays of regexes. When loading the current array from a file and iterating the entries, if the entry is a string, it applies to all and is added to the array for the primary mirror and any repos. If the entry is itself an array (tuple), the first element is the repo id, the second element is the regex.

So if the persisted current blacklist is:
`[regex1, regex2]`

and I have 1 repo, it becomes:

```
{
  None: [regex1, regex2],
  0: [regex1, regex]
}
```

When persisting, convert back to an array of tuples. When using the blacklist, fetch the dict and key by the repo id (or None for the primary mirror).

When persisting it becomes:
`[[None, regex1], [None, regex2], [0, regex1], [0, regex2]]`

Now adding and removing from the blacklist is somewhat problematic since the operations currently take a list of ids. Changing this to have the first element be the repo id would break the existing behaviour. I was thinking to leave this command which removes the blacklist from main mirror and adding two new commands `repo-blacklist-add` and `repo-blacklist-del` which take the repo id as the first argument following by the list of index values to remove.

The print-blacklist would remain the same, but would iterate all printing each blacklist in a loop (different example than above):

```
# slackroll print-blacklist
Blacklisted expressions:
    0     regex
End of list
Blacklisted expressions for repo 1:
    0     regex1
    1     regex2
End of list
Blacklisted expressions for repo 3:
    0     regexA
    1     regexB
End of list
```

### PR DESCRIPTION
**Pull request by aclemons**
Closes #8

### CODE REVIEW

File: helpfiles/blacklist-add.txt [hunk 1]
```diff
@@ -35,6 +35,11 @@ version you want to install when there's an upgrade. For example, if you don't
 use the "nosmp" kernels, you can blacklist "/extra/linux-[^-/]+-nosmp-sdk/" to
 ignore those packages.
 
+As mentioned above, the regular expressions will be used on the whole package
+name. You can additionally have a regular expression apply only to certain
+urls by adding '@' followed by a regular expression which will be used on the
```
Comment (rg3) [inline helpfiles/blacklist-add.txt]: I think we should be using "URLs" instead of "urls" in the documentation.

File: helpfiles/blacklist-add.txt [hunk 2]
```diff
@@ -35,6 +35,11 @@ version you want to install when there's an upgrade. For example, if you don't
...
+whole url.
```
Comment (rg3) [inline helpfiles/blacklist-add.txt]: Same here.

File: slackroll [hunk 3]
```diff
@@ -455,17 +455,34 @@ def get_blacklist(): # Get blacklist expression list
     return []
 
 def get_blacklist_re(): # Almost the same, but already converted to regular expressions
+    regex_list = []
```
Comment (rg3) [inline slackroll]: While reviewing this PR, I realized the empty string is a valid regular expression that always matches any string. I think we can take advantage of that to simplify the code a bit more and avoid the few if/else blocks. Let's suppose we have the following function:

    def split_blacklist_str(blacklist_str):
        if '@' not in blacklist_str:
            return (blacklist_str, '')
        return tuple(blacklist_str.split('@', 1))

You can then use that to get the pkg_regex and the url_regex inconditionally, and append the tuple without any inconvenience. If the URL part is empty, the empty regular expression will match every URL.

File: slackroll [hunk 4]
```diff
@@ -455,17 +455,34 @@ def get_blacklist(): # Get blacklist expression list
...
     except re.error:
         sys.exit('ERROR: invalid regular expression found in blacklist')
 
 def add_blacklist_exprs(expressions): # Add expressions to the blacklist
     valid_ones = []
     for regex_str in expressions:
         try:
-            re.compile(regex_str)
-            valid_ones.append(regex_str)
+            data = regex_str.split('@', 1)
```
Comment (rg3) [inline slackroll]: After splitting, we would always do the "if" block inconditionally.

File: slackroll [hunk 5]
```diff
@@ -570,7 +587,12 @@ def extend_remote_list(local_filelist, remote_list, pkgs_url=None):
     for pkg in get_remote_pkgs(local_filelist, pkgs_url):
         # Ignore packages matching blacklist
         fullname = pkg.fullname
-        matches = [x.search(fullname) for x in bl]
+        url = pkg.url(pkgs_url)
+        matches = []
+        for (pkg_regex, url_regex) in bl:
```
Comment (rg3) [inline slackroll]: You can replace this for loop and the following if with something simpler in Python. Comment added for clarity. The line's a bit long, though.

    # If the package full name and its URL match any entry in the black list
    if any((pkg_regex.search(fullname) is not None and url_regex.search(url) is not None) for (pkg_regex, url_regex) in bl:
        continue

**Review by rg3 — CHANGES_REQUESTED** [review]

File: helpfiles/blacklist-add.txt (hunk 1 above)
Comment (aclemons) [reply helpfiles/blacklist-add.txt]: Updated.

File: helpfiles/blacklist-add.txt (hunk 2 above)
Comment (aclemons) [reply helpfiles/blacklist-add.txt]: Here too.

File: slackroll (hunk 3 above)
Comment (aclemons) [reply slackroll]: Ok, I've added this. Simplifies the rest :+1:

File: slackroll (hunk 4 above)
Comment (aclemons) [reply slackroll]: Updated

File: slackroll (hunk 5 above)
Comment (aclemons) [reply slackroll]: And here

**Comment by rg3** [comment]
Merged. Thanks!

**Pull request merged**